src/stable_marriage.rs

```diff
@@ -0,0 +1,164 @@
+struct Individual {
+    cnt_match: Option<usize>,
+
+    // For efficiency: most preferred last for men, most preferred first for women
+    is_man: bool,
+    prefs: Vec<usize>,
+}
+
+impl Individual {
+    fn get_pref_idx(&self, idx: usize) -> Option<usize> {
+        if self.is_man {
+            self.prefs.iter().rev().position(|i| *i == idx)
+        } else {
+            self.prefs.iter().position(|i| *i == idx)
+        }
+    }
+
+    fn pop_preferred_if_unmatched(&mut self) -> Option<usize> {
+        if self.cnt_match.is_some() {
+            None
+        } else {
+            self.prefs.pop()
+        }
+    }
+}
+
+fn make_empty_matching(v: Vec<Vec<usize>>, is_man: bool) -> Vec<Individual> {
+    v.into_iter()
+        .map(|prefs|
+            Individual {
+                cnt_match: None,
+                is_man: is_man,
+                prefs: prefs,
+            }
+        )
+        .collect()
+}
+
+fn init_matching(
+    men: Vec<Vec<usize>>,
+    women: Vec<Vec<usize>>,
+) -> (Vec<Individual>, Vec<Individual>) {
+    let mut men = make_empty_matching(men, true);
+    let women = make_empty_matching(women, false);
+    for mut x in &mut men {
+        x.prefs.reverse();
+    }
+
+    (men, women)
+}
+
+fn extract_matching(v: &Vec<Individual>) -> Vec<Option<usize>> {
+    v.into_iter().map(|x| x.cnt_match).collect()
+}
+
+fn saturate_matching(
+    men: &mut Vec<Individual>,
+    women: &mut Vec<Individual>,
+) {
+    while let Some(i_man) = men.iter()
+        .position(|m| m.cnt_match.is_none() && !m.prefs.is_empty())
+    {
+        while let Some(i_woman) = men[i_man].pop_preferred_if_unmatched() {
+            let woman = &mut women[i_woman];
+
+            if let Some(i_man_in_w_prefs) = woman.get_pref_idx(i_man) {
+                if let Some(i_otherman) = woman.cnt_match {
+                    men[i_otherman].cnt_match = None;
```

Comment:
I feel like you're missing the `if w prefers m to m'` step of https://en.wikipedia.org/wiki/Stable_marriage_problem#Algorithm, did I miss the place where it's done?

Reply:
The woman's current preference list only contains the men that she prefers to her current match, because we update it every time that woman gets matched. So it is implicitely taken into account

Reply:
Oh indeed that was in the `.truncate`, got it!